File: tests/speed01.cpp
#include <log4cxx/logger.h>
#include "c.h"
#include <fstream>
#include <sstream>
#include <xmlbeansxx/XmlParser.h>
#include <xmlbeansxx/TextUtils.h>

using namespace xmlbeansxx;
using namespace com::p4::mind::mytest;
using namespace std;

log4cxx::LoggerPtr LOG = log4cxx::Logger::getLogger("speed01");

double currentTime() {
    struct timeval tv;
    if (gettimeofday(&tv,NULL) != 0) throw IllegalStateException();
    return double(tv.tv_sec) + double(tv.tv_usec) / 1000000.0;
}

int main() {
    //Parsing speed test
    try {
        XmlParser parser(XmlParser::create());
        parser->loadGrammar("c.xsd");
        parser->getXmlOptions()->setValidation(false);
        ifstream inStream("d.xml");
        String in = TextUtils::istreamToString(inStream);
    
        while (true) {
            ContentDocument doc = ContentDocument::Factory::newInstance();
            double t1 = currentTime();
            parser->parse(in, doc);
            double t2 = currentTime();
            string xml;
            xml = doc->toString();
            double t3 = currentTime();
            int car;
            car = cast<Fullpersoninfo>(doc->getContent()->getEmployee())->getCar();
            
            LOG4CXX_INFO(LOG, "Parse duration: " << (t2-t1) << " sec." << " " << 1. / (t2-t1) << " per second.");
            LOG4CXX_INFO(LOG, "Serialize duration: " << (t3-t2) << " sec." << " " << 1. / (t3-t2) << " per second.");
            LOG4CXX_INFO(LOG, "Car: " << car);
            //LOG4CXX_INFO(LOG, "Xml: " << xml);
        }
    } catch (xmlbeansxx::BeansException &e) {
        LOG4CXX_ERROR(LOG, "Exception: " + e.getMessage());
        throw e;
    }
    return 0;
}

File: tests/speed01.java
import org.apache.xmlbeans.*;
import com.p4.mind.mytest.*;
import java.io.*;
import org.apache.commons.logging.*;

public class speed01 {
    private static Log LOG = LogFactory.getLog(speed01.class);
    private static double currentTime() {
        return System.currentTimeMillis() / 1000.0;
    }

    public static String istreamToString(InputStream in) throws Exception {
        StringWriter out = new StringWriter();
        int c;

        while ((c = in.read()) != -1)
           out.write(c);

        in.close();
        out.close();
        return out.toString();
    }

    public static void main(String[] args) throws Exception {
        //Parsing speed test
        try {
            String in = istreamToString(new FileInputStream("d.xml"));
            while (true) {
                double t1 = currentTime();
                ContentDocument doc = ContentDocument.Factory.parse(in);
                doc.validate();
                
                double t2 = currentTime();
                
                int car = 0;
                for (int i = 0; i < 10; i++) {
                    car = ((Fullpersoninfo)(doc.getContent().getEmployee())).getCar();
                    ((Fullpersoninfo)(doc.getContent().getEmployee())).setCar(car + 1);
                }

                {
                    TreeType tree = ((Fullpersoninfo)(doc.getContent().getEmployee())).getTree().getLeft().getLeft();
                    ((Fullpersoninfo)(doc.getContent().getEmployee())).getTree().getRight().getLeft().setRight(tree);
                }
                
                double t3 = currentTime();
                
                String xml;
                {
                    StringWriter out = new StringWriter();
                    doc.save(out, new XmlOptions().setSavePrettyPrintIndent(0));
                    xml = out.toString();
                }
                double t4 = currentTime();

                LOG.info("Parse duration: " + (t2-t1) + " sec. " + (1. / (t2-t1)) + " per second.");
                LOG.info("Modify duration: " + (t3-t2) + " sec. " + (1. / (t3-t2)) + " per second.");
                LOG.info("Serialize duration: " + (t4-t3) + " sec. " + (1. / (t4-t3)) + " per second.");
                LOG.info("Car: " + car);
                //LOG.info("Xml: " + xml);
            }
        } catch (Exception e) {
            LOG.error("Exception: " + e.getMessage());
            throw e;
        }
    }
}

